<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>problems</title>
    <meta charset="utf-8" />
    <script src="p6_files/header-attrs/header-attrs.js"></script>
    <link href="p6_files/remark-css/default.css" rel="stylesheet" />
    <link href="p6_files/remark-css/default-fonts.css" rel="stylesheet" />
    <link rel="stylesheet" href="css/xaringan.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">


class: inverse, middle, center




## .LARGE[.orange[Testing your knowledge]]

---

## .Large[Problem 1]

&lt;div class="my-footer1" style="text-align: center;background-color:#D9D9D9;"&gt; .grey[Background] &lt;/div&gt;
&lt;div class="my-footer2" style="text-align: center;background-color:#D9D9D9;border-left:1px solid black;"&gt; .grey[`Bash`] &lt;/div&gt;
&lt;div class="my-footer3" style="text-align: center;background-color:#D9D9D9;border-left:1px solid black;"&gt; .grey[Scripting] &lt;/div&gt;
&lt;div class="my-footer4" style="text-align: center;background-color:#D9D9D9;border-left:1px solid black;"&gt; .grey[Permissions] &lt;/div&gt;
&lt;div class="my-footer5" style="text-align: center;background-color:#FEE0D2;border-left:1px solid black;"&gt; .blue[&lt;b&gt;Exercise&lt;/b&gt;] &lt;/div&gt;


.large[
PFAM is a domain database. It also includes domains that have already been computed for all proteomes. The most recent version is available here: [https://ftp.ebi.ac.uk/pub/databases/Pfam/releases/Pfam35.0/proteomes/](https://ftp.ebi.ac.uk/pub/databases/Pfam/releases/Pfam35.0/proteomes/). Each file represents a proteome, which is recognized by its taxonomy ID. The ID `9606` belongs to a human. Each of these files is tab-delimited, with the domain ID in the sixth column. Using `wget`, download the human proteome file. After downloading, run a single line of `bash` commands to determine the number of domain types (unique domains) in the human genome. You may use as many commands as you want, connected together in `pipes`.
]

--

&lt;br&gt;
```perl
wget https://ftp.ebi.ac.uk/pub/databases/Pfam/releases/Pfam35.0/proteomes/9606.tsv.gz
```
--
```perl
zcat 9606.tsv.gz | cut -f 6 | grep -v "#" | sort| uniq | wc -l
```
--
&lt;br&gt;
```perl
6680
```

---

## .Large[Problem 2]

&lt;div class="my-footer1" style="text-align: center;background-color:#D9D9D9;"&gt; .grey[Background] &lt;/div&gt;
&lt;div class="my-footer2" style="text-align: center;background-color:#D9D9D9;border-left:1px solid black;"&gt; .grey[`Bash`] &lt;/div&gt;
&lt;div class="my-footer3" style="text-align: center;background-color:#D9D9D9;border-left:1px solid black;"&gt; .grey[Scripting] &lt;/div&gt;
&lt;div class="my-footer4" style="text-align: center;background-color:#D9D9D9;border-left:1px solid black;"&gt; .grey[Permissions] &lt;/div&gt;
&lt;div class="my-footer5" style="text-align: center;background-color:#FEE0D2;border-left:1px solid black;"&gt; .blue[&lt;b&gt;Exercise&lt;/b&gt;] &lt;/div&gt;

.large[
Download the covid19 fasta file from [https://ftp.ebi.ac.uk/pub/databases/Pfam/releases/Pfam_SARS-CoV-2_2.0/](https://ftp.ebi.ac.uk/pub/databases/Pfam/releases/Pfam_SARS-CoV-2_2.0/) using `wget` command.

1. How many proteins are there?
2. How many amino acids are there?
3. Take only the protein sequences, and convert them to lower case.
]

--
&lt;br&gt;

```perl
cat covid-19.fasta | grep "&gt;" | wc -l
```

--

```perl
cat covid-19.fasta | grep -v "&gt;" | wc -c
```

--

```perl
cat covid-19.fasta | grep -v "&gt;" | tr [:upper:] [:lower:]
```

---

## .Large[Problem 3]

&lt;div class="my-footer1" style="text-align: center;background-color:#D9D9D9;"&gt; .grey[Background] &lt;/div&gt;
&lt;div class="my-footer2" style="text-align: center;background-color:#D9D9D9;border-left:1px solid black;"&gt; .grey[`Bash`] &lt;/div&gt;
&lt;div class="my-footer3" style="text-align: center;background-color:#D9D9D9;border-left:1px solid black;"&gt; .grey[Scripting] &lt;/div&gt;
&lt;div class="my-footer4" style="text-align: center;background-color:#D9D9D9;border-left:1px solid black;"&gt; .grey[Permissions] &lt;/div&gt;
&lt;div class="my-footer5" style="text-align: center;background-color:#FEE0D2;border-left:1px solid black;"&gt; .blue[&lt;b&gt;Exercise&lt;/b&gt;] &lt;/div&gt;

.large[
PFAM is a domain database. It also includes domains that have already been computed for all proteomes. The most recent version is available here: [https://ftp.ebi.ac.uk/pub/databases/Pfam/releases/Pfam35.0/proteomes/](https://ftp.ebi.ac.uk/pub/databases/Pfam/releases/Pfam35.0/proteomes/). Each file represents a proteome, which is recognized by its taxonomy ID. The ID `9606` belongs to a human. Each of these files is tab-delimited, with the domain ID in the sixth column.

1. Make a directory in your home folder `software/bin`.
2. Change directory to `software/bin`.
3. Write a `bash` script that take an input: taxonomy ID.
4. If file does not exist, Script should download the proteome file for the input taxonomy ID.
5. Script should run a single line of `bash` commands to determine the number of domain types (unique domains) in the genome.
6. Run step 5 **only if** step 4 is finished.
7. Save your file as `count_uniq_domains`
8. Add your script to `PATH`
9. Go to your home folder.
10. Create a `bash` script that will run `count_uniq_domains` (without specifying path). This script should store `STDOUT` and `STDERR`.
]

---

## .Large[Problem 3]

&lt;div class="my-footer1" style="text-align: center;background-color:#D9D9D9;"&gt; .grey[Background] &lt;/div&gt;
&lt;div class="my-footer2" style="text-align: center;background-color:#D9D9D9;border-left:1px solid black;"&gt; .grey[`Bash`] &lt;/div&gt;
&lt;div class="my-footer3" style="text-align: center;background-color:#D9D9D9;border-left:1px solid black;"&gt; .grey[Scripting] &lt;/div&gt;
&lt;div class="my-footer4" style="text-align: center;background-color:#D9D9D9;border-left:1px solid black;"&gt; .grey[Permissions] &lt;/div&gt;
&lt;div class="my-footer5" style="text-align: center;background-color:#FEE0D2;border-left:1px solid black;"&gt; .blue[&lt;b&gt;Exercise&lt;/b&gt;] &lt;/div&gt;


```bash
# 1 and 2

cd ~/
mkdir software
cd software
mkdir bin
```
---

## .Large[Problem 3]


```bash
# 3, 4, 5, 6, and 7

nano count_uniq_domains

#!/bin/bash

txid=$1
txid2=$txid.final.tsv.gz

if [ -e $txid2 ]
        then
zcat $txid2 | grep -v "#" | cut -f 6 | sort| uniq| wc -l
        else
wget https://ftp.ebi.ac.uk/pub/databases/Pfam/releases/Pfam35.0/proteomes/$txid.tsv.gz &amp;&amp;\
 mv $txid.tsv.gz $txid2 &amp;&amp;\
 zcat $txid2 | grep -v "#" | cut -f 6 | sort| uniq| wc -l
fi
```

---

## .Large[Problem 3]

&lt;div class="my-footer1" style="text-align: center;background-color:#D9D9D9;"&gt; .grey[Background] &lt;/div&gt;
&lt;div class="my-footer2" style="text-align: center;background-color:#D9D9D9;border-left:1px solid black;"&gt; .grey[`Bash`] &lt;/div&gt;
&lt;div class="my-footer3" style="text-align: center;background-color:#D9D9D9;border-left:1px solid black;"&gt; .grey[Scripting] &lt;/div&gt;
&lt;div class="my-footer4" style="text-align: center;background-color:#D9D9D9;border-left:1px solid black;"&gt; .grey[Permissions] &lt;/div&gt;
&lt;div class="my-footer5" style="text-align: center;background-color:#FEE0D2;border-left:1px solid black;"&gt; .blue[&lt;b&gt;Exercise&lt;/b&gt;] &lt;/div&gt;

```bash
# 8

chmod +x count_uniq_domains

export PATH="~/software/bin/:$PATH"
```

```bash
# 9

cd ~/
```

```bash
# 10

nano run.sh

#!/bin/bash

count_uniq_domains 9696 &gt; STDOUT 2&gt;STDERR

## close the file with ctrl + x

bash run.sh
```

---

## .Large[Resources]

https://github.com/urppeia/BIO609_2016

https://github.com/urppeia/BIO609_2017

https://github.com/urppeia/BIO609_2018

https://github.com/urppeia/bio609

https://github.com/urppeia/BIO609_2021

---

class: center,hide_logo

background-image: url("figs/bitly.png")
background-size: 40%
background-position: 50% 40%

## .Large[Instructor and Course evaluation]

&lt;br&gt;&lt;br&gt;&lt;br&gt;&lt;br&gt;&lt;br&gt;&lt;br&gt;&lt;br&gt;&lt;br&gt;&lt;br&gt;&lt;br&gt;&lt;br&gt;&lt;br&gt;&lt;br&gt;&lt;br&gt;&lt;br&gt;&lt;br&gt;&lt;br&gt;&lt;br&gt;&lt;br&gt;&lt;br&gt;
.content-box-green[.blue[.LARGE[[bit.ly/bio609_2024](https://bit.ly/bio609_2024)]]]
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script src="addons/macros.js"></script>
<script src="https://platform.twitter.com/widgets.js"></script>
<script>var slideshow = remark.create({
"ratio": "16:9",
"highlightLines": true,
"highlightSpans": false,
"highlightStyle": "github",
"countIncrementalSlides": false,
"slideNumberFormat": "%current%",
"navigation": {
"scroll": false
}
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
// add `data-at-shortcutkeys` attribute to <body> to resolve conflicts with JAWS
// screen reader (see PR #262)
(function(d) {
  let res = {};
  d.querySelectorAll('.remark-help-content table tr').forEach(tr => {
    const t = tr.querySelector('td:nth-child(2)').innerText;
    tr.querySelectorAll('td:first-child .key').forEach(key => {
      const k = key.innerText;
      if (/^[a-z]$/.test(k)) res[k] = t;  // must be a single letter (key)
    });
  });
  d.body.setAttribute('data-at-shortcutkeys', JSON.stringify(res));
})(document);
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
